crates/bevy_ecs/src/reflect/bundle.rs

```diff
@@ -124,49 +114,53 @@ impl ReflectBundle {
     }
 }
 
-impl<B: Bundle + Reflect + FromWorld> FromType<B> for ReflectBundle {
+impl<B: Bundle + Reflect + FromReflect> FromType<B> for ReflectBundle {
     fn from_type() -> Self {
         ReflectBundle(ReflectBundleFns {
-            from_world: |world| Box::new(B::from_world(world)),
             insert: |entity, reflected_bundle| {
-                let mut bundle = entity.world_scope(|world| B::from_world(world));
-                bundle.apply(reflected_bundle);
+                let bundle = B::from_reflect(reflected_bundle).unwrap();
                 entity.insert(bundle);
             },
             apply: |entity, reflected_bundle, registry| {
-                let mut bundle = entity.world_scope(|world| B::from_world(world));
-                bundle.apply(reflected_bundle);
-
-                match bundle.reflect_ref() {
-                    ReflectRef::Struct(bundle) => bundle
-                        .iter_fields()
-                        .for_each(|field| insert_field::<B>(entity, field, registry)),
-                    ReflectRef::Tuple(bundle) => bundle
-                        .iter_fields()
-                        .for_each(|field| insert_field::<B>(entity, field, registry)),
-                    _ => panic!(
-                        "expected bundle `{}` to be named struct or tuple",
-                        // FIXME: once we have unique reflect, use `TypePath`.
-                        std::any::type_name::<B>(),
-                    ),
+                if let Some(reflect_component) =
+                    registry.get_type_data::<ReflectComponent>(TypeId::of::<B>())
+                {
+                    reflect_component.apply(entity, reflected_bundle);
+                } else {
```

Comment:
While reworking `insert_field` I realized it would make sense to check if the `Bundle` being inserted is a single `Component`, in which case it shouldn't insert each field as if it was itself a bundle/component (which would be wrong). I guess usually `ReflectComponent` should be checked before `ReflectBundle`, but if someone made a mistake we should catch it. This won't work though if someone only registers `ReflectBundle` and not `ReflectComponent`. It's also a technically a change in behaviour (though I would argue any case where a change is observable was likely behaving wrongly before).